crates/bifrost/src/providers/replicated_loglet/replication/checker.rs

```diff
@@ -39,327 +98,747 @@ use restate_types::PlainNodeId;
+#[derive(Debug, Clone)]
+struct HashPtr<Attr>(*mut FailureDomainState<Attr>);
+
+// **Safety**: HashPtr is safe to send as it's strictly private to NodeSetChecker and is tied to
+// a heap-allocation that's also self-owned.
+unsafe impl<Attr: Send> Send for HashPtr<Attr> {}
+
+impl<Attr> HashPtr<Attr> {
+    fn as_ref(&self) -> &FailureDomainState<Attr> {
+        // Safety: We are the only one who can create HashingPtr and we guarantee that the
+        // underlying allocation is valid.
+        unsafe { &*self.0 }
+    }
+    fn as_mut_ref(&mut self) -> &mut FailureDomainState<Attr> {
+        // Safety: We are the only one who can create HashingPtr and we guarantee that the
+        // underlying allocation is valid.
+        unsafe { &mut *self.0 }
+    }
+}
+
+impl<Attr> Hash for HashPtr<Attr> {
+    fn hash<H: Hasher>(&self, hasher: &mut H) {
+        self.0.hash(hasher)
+    }
+}
+
+impl<Attr> Eq for HashPtr<Attr> {}
+impl<Attr> PartialEq for HashPtr<Attr> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl<Attr> HashPtr<Attr> {
+    fn new(value: &mut Box<FailureDomainState<Attr>>) -> Self {
+        // Critical: We want the address of the underlying allocation, not the box itself so the
+        // double dereferencing is essential.
+        //
+        // We could have made new() accept &mut FailureDomainState<Attr> but it would have made it
+        // easy to pass non-boxed references which will move.
+        let value: *mut _ = &mut **value;
+        Self(value)
+    }
+}
+
+#[derive(Debug)]
+struct LocationScopeState<Attr> {
+    /// domains at that scope
+    failure_domains: HashMap<SmartString, Box<FailureDomainState<Attr>>, Xxh3Builder>,
+    /// maps between node-id to the domain it belongs to for fast lookups
+    node_to_fd: HashMap<PlainNodeId, HashPtr<Attr>, Xxh3Builder>,
+    /// replication factor at that scope
+    replication_factor: u8,
+    /// For each attribute, the number of domains at that scope that have all their nodes with the
+    /// attribute.
+    num_complete_domains: HashMap<Attr, u32>,
+    /// For each Attribute, set of the domains at that scope that have at least one node with the
+    /// attribute. Those are the failure domains considered for
+    /// replication/write-quorum/write-quorum.
+    replication_fds: HashMap<Attr, HashSet<HashPtr<Attr>, Xxh3Builder>>,
+}
+
+impl<Attr: Eq + Hash + Clone + std::fmt::Debug> LocationScopeState<Attr> {
+    /// Checks write-quorum at this specific scope
+    fn check_write_quorum<Predicate>(&self, predicate: &Predicate) -> bool
+    where
+        Predicate: Fn(&Attr) -> bool,
+    {
+        // For each attribute, merge the set of FailureDomainState that have at least one node
+        // with the attribute here. We stop merging immediately after the set's size reaches
+        // the replication target. This makes `check_write_quorum` O(replication * num_domain_matches * num_scopes) in the
```

Comment:
This O notation does not refer to this `check_write_quorum` but the one belonging to the `NodeSetChecker`, does it? If so, maybe move it there. The num_scopes is a bit confusing in this context.

Probably it's already to late for my brain. I don't see how `replication * num_domain_matches * num_scopes` is the worst case. Isn't it a bit better because of the early break condition: `O(num_scopes * (#attributes + replication))` (worst case for every scope we need to go over all attributes and in the last attribute our predicate matches and we find `replication` domains)?

Reply:
That indeed refers to the cost that also include constructing the NodeSetChecker. I'll remove the comment to avoid confusions.